crates/rocksdb/src/db_manager.rs

```diff
@@ -0,0 +1,330 @@
+            updateable_common_opts: Box::new(updateable_common_opts),
+            current_common_opts: prev_opts,
+            subscriptions: Vec::new(),
+        };
+
+        let shutdown_watch = cancellation_watcher();
+        tokio::pin!(shutdown_watch);
+
+        let config_watch = Configuration::watcher();
+        tokio::pin!(config_watch);
+
+        loop {
+            tokio::select! {
+                biased;
+                _ = &mut shutdown_watch => {
+                    // Shutdown requested.
+                    break;
+                }
+                Some(cmd) = watchdog.watchdog_rx.recv() => {
+                    watchdog.handle_command(cmd)
+                }
+                _ = config_watch.changed() => {
+                    watchdog.on_config_update();
+                }
+            }
+        }
+
+        watchdog.shutdown();
+        Ok(())
+    }
+
+    fn shutdown(&mut self) {
+        // Ask all databases to shutdown cleanly.
+        let start = Instant::now();
+        for v in &self.manager.dbs {
+            info!(
+                db = %v.name,
+                owner = %v.owner,
+                "Shutting down rocksdb database"
+            );
+            if let Err(e) = v.db.flush_wal(true) {
+                warn!(
+                    db = %v.name,
+                    owner = %v.owner,
+                    "Failed to flush local loglet rocksdb WAL: {}",
+                    e
+                );
+            }
+            if let Err(e) = v.db.flush_memtables_atomically(v.cfs().as_slice(), true) {
+                warn!(
+                    db = %v.name,
+                    owner = %v.owner,
+                    "Failed to flush memtables: {}",
+                    e
+                );
+            }
+            v.db.cancel_all_background_work(true);
+        }
+        info!("Clean rocksdb shutdown took {:?}", start.elapsed());
+    }
+
+    fn handle_command(&mut self, cmd: WatchdogCommand) {
+        match cmd {
+            WatchdogCommand::ResetAll => {
+                self.shutdown();
+                self.manager.dbs.clear();
+                self.subscriptions.clear();
+            }
+            WatchdogCommand::Register(sub) => self.subscriptions.push(sub),
+        }
+    }
+
+    fn on_config_update(&mut self) {
+        // Memory budget changed?
+        let new_common_opts = self.updateable_common_opts.load();
+        if new_common_opts.rocksdb_total_memory_limit
+            != self.current_common_opts.rocksdb_total_memory_limit
+        {
+            info!(
+                old = self.current_common_opts.rocksdb_total_memory_limit,
+                new = new_common_opts.rocksdb_total_memory_limit,
+                "[config update] Setting rocksdb total memory limit to {}",
+                ByteSize::b(new_common_opts.rocksdb_total_memory_limit)
+            );
+            self.cache
+                .set_capacity(new_common_opts.rocksdb_total_memory_limit as usize);
```

Comment:
I assume that the cache will free up resources if the new limit is lower than the current usage, right?